akka-actor/src/main/scala/akka/pattern/RetrySupport.scala

```diff
@@ -36,6 +36,32 @@ trait RetrySupport {
     RetrySupport.retry(attempt, attempts, attempted = 0)
   }
 
+  /**
+   * Given a function from Unit to Future, returns an internally retrying Future.
+   * The first attempt will be made immediately, each subsequent attempt will be made immediately
+   * if the preceding attempt failed and the 'shouldRetry' predicate, when applied to the failure's
+   * exception evaluates to 'true'.  If the predicate evaluates to 'false', the failed attempt will
+   * be returned.  This allows for short-circuiting in situations where the retries cannot be expected
+   * to succeed (e.g. in a situation where the legality of arguments can only be determined asynchronously).
+   *
+   * If attempts are exhausted, the returned future is the that of the last attempt.
+   * Note that the attempt function will be invoked on the given execution context for subsequent tries and
+   * therefore must be thread safe (not touch unsafe mutable state).
+   *
+   * <b>Example usage</b>
+   *
+   * {{{
+   * def possiblyFailing(): Future[Something] = ???
+   * val withRetry: Future[Something] = retry(
+   *   attempt = possiblyFailing,
+   *   attempts = 10,
+   *   shouldRetry = { (ex) => ex.isInstanceOf[IllegalArgumentException] })
+   * }}}
+   */
+  def retry[T](attempt: () => Future[T], shouldRetry: Throwable => Boolean, attempts: Int)(
```

Comment:
Align order of parameters with the other signature with more params, either attempts in between in both or lambdas first in both? (Unless that causes compiler ambiguity, but I don't think it should)

Reply:
`attempts` in-between does cause compiler confusion (with the `delayFunction` variants), so would break source compatibility, but `shouldRetry` after `attempt` does work.